### DISCUSSION TITLE
Loading an ARMI DB when you don't have the App that created it

### DISCUSSION THREAD

**Discussion by john-science** [terrapower/armi#1688] [General]
Right now, the story goes that you can't load an ARMI DB if you don't have the ARMI `App` that was used to create the DB in the first place.

This displeases me.

It means a lot more overhead in sharing ARMI outputs. You need to go find the `App` in question, get all its data and cruft, just to load an H5 file. ARMI should be easier to use than that. 

So, the question is: Why do you need the `App`?

* You need the `App` because that's where the `Parameter`s in the Database are defined.
* You need the `App` because it defines the `Setting`s in the `cs` object.

Is there any other reason?

**Comment by john-science** [comment]
@jakehader @keckler @onufer I would love you input here.  Is it _just_ the `Parameter` / `Settings` definitions that are the issue?

This is a "I don't know what I don't know" situation.

Thanks!

**Reply by jakehader** [reply]
There could be other objects that are being created, like special subclassed assemblies, blocks, or components, or even material data where the application can register these. 

If it was just parameters and settings then I think ignoring those for a database load could be reasonable. I do think there is an opportunity to figure out how we may get data off of the data model if there isn't a complete application to recreate it but we might need to understand this a bit better when it comes to trying to use the data model to integrate values.

Some type of lightweight placeholders that maintain key:value pairs in the data model may be achievable and could make database loading a lot faster if we make some assumptions.

**Reply by john-science** [reply]
> Some type of lightweight placeholders that maintain key:value pairs in the data model...

I like this. Okay, here are three ideas in decreasing order of severity and complexity:

1. Enforce that classes like custom `Settings`, `Parameters`, `SpecialBlocks`, or `CustomMaterials` are fully saved in the DB, and we can use them to reload a DB.
2. Maintain some lightweight version of each of the custom things that are loading into the DB, so we can load, and _maybe_ run from an existing DB.
3. Allow users to load a DB in some sort of "Safe Mode" where, sure, you can't DO a run with it, because you don't know what the hell a given `CustomMaterial` or `CustomParameter` is.  But at least you can LOOK at everything by name. And inspect the DB/reactor.

**Reply by jakehader** [reply]
I think there could be value in having a reactor data model in safe mode more broadly as well! I'd love the capability to lock the state from modification even as we are developing an `Interface` to protect against unintentional changes if the code being implemented isn't supposed to change the model. I'm imagining the the data model is in safe mode by default and there is some context manager like `retainState` that you need to enter to assign or modify parameters or state variables. Maybe an application can establish this as a rule for the operator and users can see how "safe" they need or would like to be.

I am thinking about this because to me, having an `Interface` prescribe how and when it will make reactor modifications could help us better enforce and test the scope of code. 

As for storing special information, we may want to start with requirements for loading a database. I'm not sure what storing a custom material would look like and if this would be storing the custom code or making material parameters to use the parameter system.

**Reply by jeffbaylor** [reply]
I really like this idea.

If the database carried sufficient information to reproduce all the `Settings` and `Parameter` definitions for each type of `Parameter` data it contains, I think it would be a good start and maybe sufficient.  There would still be the burden on the loading `App` to have either a subset of the creator App's plugins or provide plugins that implement the interfaces for missing the objects like any custom `Composite` or `Material` objects.

This would be enough to allow databases to transfer data between ARMI apps.

**Reply by john-science** [reply]
So, @jeffbaylor, you are voting for option (1) above?

**Reply by alexhjames** [reply]
I like option 3 above. Especially if the unsupported thing was also associated with the name of the plugin that created it. So for example if you loaded a database with a material that wasn't in ARMI, you would be able to see the material name at least everywhere that you would normally see the material object and also some way to see what plugin it came from. That way you would know which plugin you needed to include if you wanted to make that a fully fleshed out object on load. 

For the custom classes that extend ARMI classes, like say a custom block class, it would be cool if the load could instance it as at least the closest thing in the inheritance tree that is available in ARMI and then only the missing pieces are just names that are stored. So for example, a custom block that inherits from a HexBlock would be instanced as a HexBlock and only the extended things would be listed as key/value pairs but it would still respond to other block method calls that are more general.

**Reply by keckler** [reply]
I would vote for (3).

**Reply by jeffbaylor** [reply]
I will have to think more about it, but my first impression was a hybrid of 1 and 3.  The `Settings` and `Parameter` objects seem well suited to be fully defined within the database and then any missing at load time could be created.  

The more complex objects like customized `Block` and `Material` objects could continue to be provided either by adding the same plugin to the loading `App`, or by a different plugin that defines the same interface to those objects, if the loading `App` wants to do anything more than view the data.  If any of those objects are not defined, then the loading `App` enters a "Safe Mode" and provides information about what missing functionality drove it to adopt this mode.  An incremental development step could be to fail if the objects are not available.

**Reply by jasonbmeng** [reply]
I like Option (3) as well. Seems like it could be fairly straightforward to implement as a starting point. More sophisticated loading functionality that has been suggested can be expanded at a later point. But in the meantime, ARMI Apps will at least be able to load and interrogate dbs generated by other ARMI Apps.

**Reply by john-science** [reply]
@jeffbaylor I believe I have a version of this that is working. Just FYI.

**Comment by keckler** [comment]
You also need for the relevant material classes to be available. If ARMI
can't find a material that is used in the reactor model, you can't do
db.load()

**Reply by john-science** [reply]
Materials.  Okay, damn.

Maybe we could have a "load DB in safe mode" that would load the DB with just the names of the materials, but not the full material data.  You would be limited with what you could do vis-a-vis, continuing the simulation from that point point.  But at least you could LOAD it.

Hmmm

**Comment by keckler** [comment]
We have an internal visualization tool (XTView) that loads DBs to some extent without having access to the original app. I'm not sure how that works, but it might be a starting point for what you're trying to accomplish here.

**Reply by ntouran** [reply]
it knows the hdf5 structure and just reads and displays data without any ARMI data model whatsoever. This means there is some duplication of geometry engines/knowledge within the two

Similarly, users can certainly open up a hdf5 file with a typical hdf5 browser and no ARMI whatsoever and read all the data out of it if you want without any changes needed to the code or DB.

**Comment by ntouran** [comment]
I feel like I'm a big fan of option 1. Storing parameter defs and settings in the DB is borderline trivial from a data magnitude perspective, and putting them in there would indeed facilitate sharing of output DBs in a way that would be more portable, e.g. into other ARMI apps or just pure framework. 

Materials do scare me a little. If you're missing materials you won't really be able to e.g. get mass density and therefore won't be able to look at compositions in any detail. Then you can't getMass or anything and many of the useful ARMI post-processing features will not work. 

Of course if you don't want any ARMI data model abstractions whatsoever you can just read the hdf5 file directly (this is what xtview does)
